sv.c

```diff
@@ -4771,6 +4776,272 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
+                /* Either it's a shared hash key, or it's suitable for
+                   copy-on-write.  */
+#ifdef DEBUGGING
+                if (DEBUG_C_TEST) {
+                    PerlIO_printf(Perl_debug_log, "Copy on write: ssv --> dsv\n");
+                    sv_dump(ssv);
+                    sv_dump(dsv);
+                }
+#endif
+#ifdef PERL_ANY_COW
+                if (!(sflags & SVf_IsCOW)) {
+                        SvIsCOW_on(ssv);
+                        CowREFCNT(ssv) = 0;
+                }
+
+                if (len) {
+                   if (sflags & SVf_IsCOW) {
+                        sv_buf_to_rw(ssv);
+                    }
+                    CowREFCNT(ssv)++;
+                    SvPV_set(dsv, SvPVX_mutable(ssv));
+                    sv_buf_to_ro(ssv);
+                } else
+#endif
+                {
+                /* SvIsCOW_shared_hash */
+                DEBUG_C(PerlIO_printf(Perl_debug_log,
+                                      "Copy on write: Sharing hash\n"));
+                SvPV_set(dsv,
+                         HEK_KEY(share_hek_hek(SvSHARED_HEK_FROM_PV(SvPVX_const(ssv)))));
+                }
+                SvLEN_set(dsv, len);
+                SvCUR_set(dsv, cur);
+                SvIsCOW_on(dsv);
+            } else {
+                /* Failed the swipe test, and we cannot do copy-on-write either.
+                   Have to copy the string.  */
+                sv_grow_fresh(dsv, cur + 1);
+
+                Move(SvPVX_const(ssv),SvPVX(dsv),cur,char);
+                SvCUR_set(dsv, cur);
+                *SvEND(dsv) = '\0';
+            }
+        SvFLAGS(dsv) |= sflags & SVf_UTF8;
+}
+
+/* A helper for sv_mortalcopy_flags and newSVsv_flags, more efficient
+ * for their use cases compared to Perl_sv_setsv_flags. */
+PERL_STATIC_INLINE SV *
+S_sv_freshcopy_flags(pTHX_ SV* ssv, const I32 flags)
+{
+    /* Note: If a PVIV/PVNV/PVMG is only IOK, NOK, ROK, it is _mostly_
+     * possible to create just a headless SV to store that value.
+     * Some parts of core (Perl_amagic_call in gv.c specifically) do
+     * assume - and possibly CPAN might - that SvTYPE(dsv) == SvTYPE(ssv)
+     * though, which is why the code below does not try that type
+     * simplification. Perhaps this might be worth revisiting in the future.
+     * -- April 2025. */
+    assert(ssv);
+    svtype stype = SvTYPE(ssv);
+
+#if NVSIZE <= IVSIZE
+    ASSUME(stype > SVt_NV);
+#else
+    ASSUME(stype > SVt_IV);
+#endif
+    ASSUME(stype <= SVt_PVMG);
+    SV *dsv = newSV_type(stype);
+    U32 sflags   = SvFLAGS(ssv);
+
+    ASSUME(!(flags & SV_GMAGIC)); /* Caller should have sorted this. */
+    /* Note: Perl_sv_setsv_flags essentially ignores magic, except for
+     * taint and vstring magic. */
+
+    /* The switch cases are in the order that gcov of make/make test suggests. */
+    switch(sflags & (SVp_IOK|SVp_NOK|SVf_ROK|SVp_POK|SVf_FAKE|SVppv_STATIC)) {
+        case SVp_IOK:
+            SvIV_set(dsv, SvIVX(ssv));
```

Comment:
SVt_IV and some build configs SVt_NV, are bodyless, please write direct to the 4th SV head field and skip indirecting through SvANY. Grep `sv.c` for ideas on how to do the fresh bodyless SV head field 4 assignment.

Reply:
The intended callers of `S_sv_freshcopy_flags` are `Perl_sv_newSV_flags` and `Perl_sv_newmortal_flags`, which at the moment are handling the bodyless cases themselves directly, so there's currently no need for `S_sv_freshcopy_flags` to also have that code.

(If that situation changes during this PR lifecycle, I will add the bodyless cases in.)

Reply:
I see the name of this fn is `S_sv_freshcopy_PVxx`, its not reachable from another libperl TU or reachable from a .dll, there you are correct, `S_sv_freshcopy_PVxx` can do whatever it wants, its private to its same TU callers.